src/commands/cmd_explain.c

```diff
@@ -6,70 +6,102 @@
+    gc->graph_name = rm_strdup("");
+
+    pthread_setspecific(_tlsGCKey, gc);
+    return gc;
+}
+
 /* Builds an execution plan but does not execute it
  * reports plan back to the client
  * Args:
- * argv[1] graph name
+ * argv[1] graph name [optional]
  * argv[2] query */
 int MGraph_Explain(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
-    if (argc < 3) return RedisModule_WrongArity(ctx);
+    if(argc < 2) return RedisModule_WrongArity(ctx);
 
     // Return immediately if invocation context is Lua or a MULTI/EXEC block
     int flags = RedisModule_GetContextFlags(ctx);
-    if (flags & (REDISMODULE_CTX_FLAGS_MULTI | REDISMODULE_CTX_FLAGS_LUA)) {
+    if(flags & (REDISMODULE_CTX_FLAGS_MULTI | REDISMODULE_CTX_FLAGS_LUA)) {
         RedisModule_ReplyWithError(ctx, "RedisGraph commands may not be called from non-blocking contexts.");
         return REDISMODULE_OK;
     }
 
-    const char *graphname = RedisModule_StringPtrLen(argv[1], NULL);
-    const char *query = RedisModule_StringPtrLen(argv[2], NULL);
+    const char *query;
+    const char *graphname = NULL;
+    bool free_graph_ctx = false;
+
+    if(argc == 2) {
+        query = RedisModule_StringPtrLen(argv[1], NULL);
+    } else {
+        graphname = RedisModule_StringPtrLen(argv[1], NULL);
+        query = RedisModule_StringPtrLen(argv[2], NULL);
+    }
 
     /* Parse query, get AST. */
-    char *errMsg = NULL;
     AST** ast = NULL;
+    char *errMsg = NULL;
     GraphContext *gc = NULL;
     ExecutionPlan *plan = NULL;
 
     ast = ParseQuery(query, strlen(query), &errMsg);
-    if (!ast) {
+    if(!ast) {
         RedisModule_Log(ctx, "debug", "Error parsing query: %s", errMsg);
         RedisModule_ReplyWithError(ctx, errMsg);
         free(errMsg);
         return REDISMODULE_OK;
     }
 
-    // Retrieve the GraphContext and acquire a read lock.    
-    gc = GraphContext_Retrieve(ctx, graphname);
-    if(!gc) {
-        RedisModule_ReplyWithError(ctx, "key doesn't contains a graph object.");
-        goto cleanup;
-    }
-
-    // Perform query validations before and after ModifyAST
-    if (AST_PerformValidations(ctx, ast) != AST_VALID) return REDISMODULE_OK;
-
+    // Perform query validations before and after ModifyAST.
+    if(AST_PerformValidations(ctx, ast) != AST_VALID) return REDISMODULE_OK;
     ModifyAST(ast);
-    if (AST_PerformValidations(ctx, ast) != AST_VALID) return REDISMODULE_OK;
+    if(AST_PerformValidations(ctx, ast) != AST_VALID) return REDISMODULE_OK;
 
-    if (ast[0]->indexNode != NULL) { // index operation
+    // index operation.
+    if(ast[0]->indexNode != NULL) {
+        RedisModule_ReplyWithArray(ctx, 1);
         char *reply = (ast[0]->indexNode->operation == CREATE_INDEX) ? "Create Index" : "Drop Index";
         RedisModule_ReplyWithSimpleString(ctx, reply);
         goto cleanup;
     }
 
+    // Retrieve the GraphContext and acquire a read lock.
+    if(graphname) {
+        gc = GraphContext_Retrieve(ctx, graphname);
```

Comment:
Why not structure it so that if this call fails, we use the `_emptyGraphContext` approach? Then we would avoid `key doesn't contains a graph object` errors on uncreated graphs.

Strictly, queries that create or do not require a pre-existing graph (CREATE/MERGE/UNWIND) shouldn't raise this error, but read queries should. I think that never raising this error is better than always raising it as we do now, though.

Reply:
I think that, in the case where the caller took the time to specify a graph ID (now optional), he/she expects the execution-plan to be constructed against specified graph, if the graph doesn't exists, for example as a result of a mistype we should let the caller know the GraphID specified is invalid.

Reply:
That's fair! We should document that the command is key-optional soon, doesn't need to be done in this PR.